### DISCUSSION TITLE
如何对每个c文件单独设置宏定义值？

### DISCUSSION THREAD

**Discussion by PeakRacing** [xmake-io/xmake#4563] [Q&A]
如何对每个c文件单独设置宏定义值？
使用before_build_file可以在每个源文件编译前处理，但是如何修改宏呢？没有接口我尝试使用
    for _, dep_define_flasg in pairs(target:get("defines")) do
        if dep_define_flasg:startswith("__CURRENT_FILE_NAME__=") then
            print("dep_define_flasg",dep_define_flasg)
            dep_define_flasg = "__CURRENT_FILE_NAME__=qqqqqqqqqqqq"
        end
    end
打印出来看还是之前的，qqqqqqqqqqqq并未生效

**Comment by star-hengxing** [comment]
```lua
add_files("test/*.c", "test2/test2.c", {defines = "TEST2", languages = "c99", includedirs = ".", cflags = "-O0"})
```

**Reply by PeakRacing** [reply]
你好，add_files这个我知道，但是我工程文件比较多，没办法手动一个一个设置，所以想before_build_file每个文件动态获取文件名，之后将宏的值设置为文件名，想要这样动态设置

**Comment by waruqi** [comment]
```lua
for _, file in ipairs(os.files(path.join(os.scriptdir(), "test/*.c"))) do
    add_files(file, {defines = "XXX"})
end
```

**Comment by PeakRacing** [comment]
> ```lua
> for _, file in ipairs(os.files(path.join(os.scriptdir(), "test/*.c"))) do
>     add_files(file, {defines = "XXX"})
> end
> ```
我们开放的是sdk，用户会直接add_files，而不是这么处理，文件和目录较多

**Reply by waruqi** [reply]
搞个 rule ，通过 target:fileconfig_set/target:fileconfig_add 配置
